### DISCUSSION TITLE
Need some clarification regarding data portal changes

### DISCUSSION THREAD

**Discussion by Brannos1970** [MarimerLLC/csla#2632] [Questions] [answered]
I am working on understanding the changes to the data portal and need some help.

Firstly, I have the following code in my BusinessBase or similarly in another Base:

`
public static async Task<CustomerEdit> GetAsync(int customerId)
{
     return await DataPortal.FetchAsync<CustomerEdit>(customerId);
}
`

How do we do this now? We no longer have a constructor, not even a private one, that can get the IDataPortal from DI. Do we need to create IDataPortal in the Task? If so how, as the ApplicationContext is not available in the static methods?

**Comment by rockfordlhotka** [comment, marked as answer]
Static factory methods are no longer an option.

Your choices are to inject (or create) a `DataPortal<T>` instance and use it, or put your static factory methods into a factory class (so they become instance methods of that class) and inject the `DataPortal<T>` into that class.

For the most part, you'll find that your UI elements (page/form/etc.) are created via DI, and so you can just inject the `DataPortal<T>` you need into the page/form and use it to interact with your domain objects.

**Reply by Brannos1970** [reply]
Just curious and I could not find any information, why did we lose the DataPortal? It seems to me that I could possibly create a static class that implements the DataPortal throughout my project. If that is correct then should we not have it in the core? I know I am missing something here. 

This change also adds a level of complexity that is requiring some major changes to an app I am updating to Csla6 which has me concerned.

**Reply by rockfordlhotka** [reply]
The data portal requires `ApplicationContext`, and `ApplicationContext` is now an instance type that is only available via DI.

`ApplicationContext`, in turn, requires access to the current service provider, which is only available via DI.

It turns out that dependency injection is much like async/await, in that once you start using it, the concept "infects" all the rest of your code.

Because there's no way to inject a value into a static method, all the static methods throughout CSLA had to be changed to instance methods - it was a massive undertaking!!

Sadly, it also means that static methods (like static factory methods) are no longer available in your code either.

You _could_ manually create an instance of `DataPortal<T>`. But to do that you'd need to create an instance of `ApplicationContext`. And to do that you'd need to acquire a reference to the current/correct `ServiceProvider` - and that's the sticking point. The only reliable way to get access to the current `ServiceProvider` is via DI.

**Reply by rockfordlhotka** [reply]
I do think, that if you can refactor your UI to rely on DI, that the changes ultimately will make your code smaller and simpler. The ability to inject necessary services into the UI (or viewmodel/controller/etc.) is extremely powerful, and really helps reduce the amount of 'plumbing' code that you need to write and maintain.

That said, I appreciate that switching to a DI-based model in an existing codebase is a really large effort. Trust me, I spent many, many, many hours doing exactly that in CSLA!

**Reply by Brannos1970** [reply]
Understood. I am working through it now. I am so used to using DataPortal that it is strange having to inject IDataPortal or IChildDataPortal for child objects into these factories. 

Btw, I am not really familiar with the factory pattern, do I create a factory for BusinessListBase objects or do I use the child object's factory?

**Reply by rockfordlhotka** [reply]
'factory' is used in different ways.

Are you referring to a factory on the client that calls the data portal, or a factory used to implement the DAL on the server?

**Reply by Brannos1970** [reply]
Factory used on the client that calls the data portal. 

I have a customer with three child objects like below. 

```c#
[Serializable]
    public class CustomerEdit : BusinessBase<CustomerEdit>
    {
        public AddressEdit MailingAddress
        {
            get => GetProperty(MailingAddressProperty);
            set => SetProperty(MailingAddressProperty, value);
        }

        public static readonly PropertyInfo<AddressEdit> PhysicalAddressProperty = RegisterProperty<AddressEdit>(nameof(PhysicalAddress));

        public AddressEdit PhysicalAddress
        {
            get => GetProperty(PhysicalAddressProperty);
            set => SetProperty(PhysicalAddressProperty, value);
        }

        public static readonly PropertyInfo<CustomerLocations> LocationsProperty = RegisterProperty<CustomerLocations>(nameof(Locations));

        public CustomerLocations Locations
        {
            get => GetProperty(LocationsProperty);
            set => SetProperty(LocationsProperty, value);
        }
}
```

I assume that for the AddressEdit object I would call 'new AddressEditFactory()' or access through IChildDataportal<AddressEdit>; but for the CustomerLocations do I create a factory for it and access through the IChildDataPortal<CustomerLocations> or do I just call the CustomerLocationsFactory? Or, do I loop through the data and call the IChildDataPOrtal<CustomerEdit>?

**Reply by rockfordlhotka** [reply]
Something like this for a simple editable root:

```c#
public class CustomerFactory
{
    private IDataPortal<CustomerEdit> CustomerPortal { get; }

    public CustomerFactory(IDataPortal<CustomerEdit> customerPortal)
    {
        CustomerPortal = customerPortal;
    }

    public async Task<CustomerEdit> GetCustomer(int id)
    {
        return await CustomerPortal.FetchAsync(id);
    }
}
```

**Reply by rockfordlhotka** [reply]
The rules for how the data portal work haven't changed - you still only use the client-side data portal to interact with root types. Child types are instantiated on the server like normal.

The reason there's an `IChildDataPortal<T>` is for use in your _server-side code_. For example, in your CustomerEdit class:

```c#
[Fetch]
private async Task Fetch(int id, [Inject]ICustomerDal dal, [Inject]IChildDataPortal<CustomerLocations> locationsPortal)
{
  var customerData = dal.GetData(id);
  var customerLocations = await locationsPortal.FetchChildAsync(id);
}
```

**Reply by Brannos1970** [reply]
Excellent, Thank you
